nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/Interval.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.web.util;
+
+public interface Interval {
+
+    enum RelativePosition {
+        /**
+         * The compared interval ends before the actual, there is no intersection.
+         */
+        BEFORE,
+
+        /**
+         * The compared interval exceeds the actual both at the low and high ends.
+         */
+        EXCEEDS,
+
+        /**
+         * The compared interval's tail (but not the whole interval) intersects the actual interval (part of it or the whole actual interval).
+         */
+        TAIL_INTERSECTS,
+
+        /**
+         * The compared interval is within the actual interval. It can match with the actual or contained by that.
+         */
+        WITHIN,
+
+        /**
+         *The compared interval's head (but not the whole interval) intersects the actual interval  (part of it or the whole actual interval).
+         */
+        HEAD_INTERSECTS,
+
+        /**
+         * The compared interval starts after the actual, there is no intersection.
+         */
+        AFTER,
+    }
+
+    /**
+     * Relative position of the "other" interval compared to this.
+     *
+     * @param otherIntervalLowerBoundary Lower boundary of the compared interval.
+     * @param otherIntervalHigherBoundary Higher boundary of the compared interval.
+     *
+     * @return Returns the relative position of the "other" interval compared to this interval. For example: if the result
+     *         is BEFORE, read it as: the other interval ends BEFORE the actual (and there is no intersection between them).
+     */
+    RelativePosition getRelativePositionOf(final int otherIntervalLowerBoundary, final int otherIntervalHigherBoundary);
+
+    /**
+     * @return Returns an interval instance with closed low and open high boundary.
+     */
+    static Interval getClosedOpenInterval(final int lowerBoundary, final int higherBoundary) {
```

Comment:
I do not agree with having a factory method in the generic interface: if further classes were to be added that implement the interface, then the interface would need to be modified each time to contain factory methods for each implementing type if we wanted to maintain consistency. I think this method could be removed and instead of calling `Interval.getClosedOpenInterval(offset, higherBoundary)` in PaginationHelper, `new ClosedOpenInterval(offset, higherBoundary)` constructor could be used.

Reply:
I understand your thought that since there is a finite number of combinations of boundaries, the interface could have methods for instantiating each combination and would not need to expose which implementing class is used. However, this current implementation violates 3 of the SOLID principles, namely:
- Single Responsibility: the interface now not only defines the capabilities of an Interval, but is also responsible for creating instances.
- Open-Closed Principle: Adding new implementations other than ClosedOpenInterval requires modification of Interval (to add new factory methods or modify exiting one(s)).
- Dependency Inversion: currently the interface depends on the implementing class.

Reply:
I slightly disagree with this primarily because of the static method usage but I have no strong feelings about it. Please expect a commit soonish